#!/bin/bash 
#PBS -N myk
#PBS -r n 
#PBS -q batch
#PBS -m a
#PBS -M marroni@appliedgenomics.org
#PBS -l nodes=1:ppn=12,vmem=150gb,walltime=168:00:00
NPROCS=$(wc -l < $PBS_NODEFILE)



#Should we hardcode directories or pass them as parameters?

DBDIR=/projects/populus/ep/share/marroni/EOSC/
READDIR=/projects/populus/ep/share/marroni/EOSC/reads/Transcriptomics
KRAKBIN=/projects/populus/ep/software/kraken2-2.0.6-beta/bin
KDB=/projects/igats/metagenomics/custom_databases/kraken_nt_2.0.7
KRES=/projects/populus/ep/share/marroni/EOSC/classification/Transcriptomics

for READFILE in $READDIR/*.gz
do
echo $READFILE
pref=$(basename ${READFILE/.fastq.gz/})
echo $pref
#This one was in case of paired, still need to deal with that
# echo "cd ${KRES}; export TMPDIR=${KRES}; module use --append /iga/scripts/dev_modules/modules; \
# ${KRAKBIN}/kraken2 --threads 16 --paired --gzip-compressed --db $KDB ${READ1} ${READ2} --output ${KRES}/${pref}.kraken --use-names --report ${KRES}/${pref}.kraken.report.txt >${KRES}/logs/map.out 2>${KRES}/logs/map.err"| qsub -N s1_${pref} -l vmem=150G,walltime=168:00:00,nodes=1:ppn=16
cd ${KRES}; export TMPDIR=${KRES}; module use --append /iga/scripts/dev_modules/modules; \
${KRAKBIN}/kraken2 --threads 16 --gzip-compressed --db $KDB ${READFILE} --output ${KRES}/${pref}.kraken \
--use-names --report ${KRES}/${pref}.kraken.report.txt >${KRES}/logs/${pref}_map.out 2>${KRES}/logs/${pref}_map.err
done
